### ISSUE TITLE
Docker-Compose throws an error while debugging in VS 2019

### ISSUE THREAD

**Issue by Urgen-Dorjee** [microsoft/DockerTools#239]
Docker file:-
> FROM mcr.microsoft.com/dotnet/core/sdk:3.1 AS build-env
WORKDIR /app
COPY *.csproj ./
RUN dotnet restore
COPY . ./
RUN dotnet publish -c Release -o out
FROM mcr.microsoft.com/dotnet/core/aspnet:3.1
COPY --from=build-env /app/out .
EXPOSE 5000/tcp
ENV ASPNETCORE_URLS http://*:5000
HEALTHCHECK --interval=30s --timeout=3s --retries=1 CMD curl --silent --fail http://localhost:5000/hc || exit 1
ENTRYPOINT ["dotnet", "VehicleManagementAPI.dll"]

Docker-compose:-
* simplified for brevity..
* please ignore yml line formatting 

> _services:
      vehiclemanagementapi:
         image: auto/vehiclemanagementapi:1.0
         build: src/VehicleManagementAPI
         depends_on:
           - rabbitmq
           - sqlserver
         ports:
            - "5000"
         environment:
            - ASPNETCORE_ENVIRONMENT=Production_

This works fine when I try to build/run the docker-compose from CLI, however this setup won't work in VS2019 for debugging purpose. It throw an exceptions :- 

`Severity	Code	Description	Project	File	Line	Suppression State
Error	CTC1000	(Line: 45, Col: 12, Idx: 955) - (Line: 45, Col: 36, Idx: 979): Exception during deserialization	docker-compose	C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Sdks\Microsoft.Docker.Sdk\build\Microsoft.VisualStudio.Docker.Compose.targets	304	
`

Therefore I had to change/setup docker file like following:-
> _FROM mcr.microsoft.com/dotnet/core/sdk:3.1 AS build-env
WORKDIR /app
FROM mcr.microsoft.com/dotnet/core/runtime:3.1
WORKDIR /src
COPY ["VehicleManagementAPI/VehicleManagementAPI.csproj", "VehicleManagementAPI/"]
COPY . .
WORKDIR /src/VehicleManagementAPI
RUN dotnet build "VehicleManagementAPI.csproj" -c Release -o /app
RUN dotnet restore -s https://api.nuget.org/v3/index.json -s https://www.myget.org/F/autoweb/api/v3/index.json
FROM build AS publish
RUN dotnet publish "VehicleManagementAPI.csproj" -c Release -o /app
HEALTHCHECK --interval=30s --timeout=3s --retries=1 CMD curl --silent --fail http://localhost:5000/hc || exit 1
FROM base AS final
WORKDIR /app
COPY --from=publish /app .
EXPOSE 5000/tcp
ENV ASPNETCORE_URLS=http://*:5000
ENTRYPOINT ["dotnet", "VehicleManagementAPI.dll"]_

Why VS2019 requires different dockerfile setup for debugging.

**Comment by dbreshears** [comment]
@Urgen-Dorjee , it does look like an issue in the tooling,  Does the following format workaround the issue for you? 

 build:
        context: src/VehicleManagementAPI

**Comment by Urgen-Dorjee** [comment]
I am curious why VS2019 won't compile/build docker file when it works fine in CLI. 
```
build:
context: src/VehicleManagementAPI
```
 this works, but why would  VS2019 needs additional line while debugging where as it works fine in CLI mode.

**Comment by dbreshears** [comment]
The VS extension builds a model around the docker-compose file and does this through a 3rd party library that aids in de-serializing of the yaml.  We do this in order to generate additional docker compose files in the obj\Docker to facilitate local debugging, https and usersecrets support, additional volume mounts, etc...

We have looked at this issue and it is a bug due to a format we just hadn't anticipated.  It has been this way for a long time and likely the beginning,  and the first that we had seen of it. 

The unfortunate part is it is not an easy fix and also a bit "risky".   Since there is an easy workaround we would I'd like to hold off on addressing unless we find this is a blocking issue

**Comment by Urgen-Dorjee** [comment]
Okay, I will leave with it as of now as you stated in your comment section. But somehow Microsoft should come up with a fix, as I had to spend considerable amount of time tying to fix it and even an error description did not really point to the right problem. I tried following step before I post my issue here:-

```
services:
  vehiclemanagementapi:
     image: auto/vehiclemanagementapi:1.0
     build: .
     depends_on:
       - rabbitmq
       - sqlserver
     ports:
        - "5000"
    environment:
        - ASPNETCORE_ENVIRONMENT=Production
```

 and this works fine, but I really wanted the way it compiles in CLI mode, therefore I had to open an issue to get the idea what is going on around docker-compose file in visual studio while debugging.

Anyway, thanks for you reply..

**Comment by tkf144** [comment]
Ran across this GH issue while trying to locate the source of a VS problem in dealing with my `docker-compose.yml` which features a `context:` value that is an environment variable provided by `.env` files, e.g.

```yaml
    build:
      context: ${DOCKERFILE_CONTEXT_PATH:?Dockerfile context path not set in .env file}
      dockerfile: ./snip/Dockerfile
```

The above-mentioned deserialization error is produced.

I'd like to use the simplest, Docker-supported mechanisms I can. I can't think of a workaround that won't require restructuring my `.env` files into compose files overrides, and avoiding `.env` files altogether.

To confirm: is support for variables via `.env` in `docker-compose.yml` lacking in the Visual Studio Docker tooling, or am I missing something?

Thank you.

**Comment by NCarlsonMSFT** [comment]
@tkf144 currently, for performance reasons, we use a slim parser during a number of debugger operations to make decisions about what services connect to what projects etc.. That parser currently doesn't process environment variables; for performance, and b/c it didn't impact most decisions. An environment variable in the build context is unfortunately a case where that does have impact. The best way to get support added would be to open a suggestion item on [Developer Community](https://developercommunity.visualstudio.com/report?entry=suggestion&space=8) explaining the scenario where you need that support. That will then be prioritized based on votes.

**Comment by NCarlsonMSFT** [comment]
@Urgen-Dorjee sorry for not following up, this was fixed in VS 2022 17.1
